util/java/libtiled-java/src/main/java/org/mapeditor/io/TMXMapReader.java

```diff
@@ -162,11 +175,15 @@ private static double getDoubleAttribute(Node node, String attribname, double de
     }
 
     private <T> T unmarshalClass(Node node, Class<T> type) throws JAXBException {
-        UnmarshallerPool unmarshallerPool = cachedUnmarshallers.get(type);
-        if (unmarshallerPool == null) {
-            JAXBContext context = JAXBContext.newInstance(type);
-            unmarshallerPool = new UnmarshallerPool(context);
-            cachedUnmarshallers.put(type, unmarshallerPool);
+        boolean typeIsBound = classesToBeBound.contains(type);
+        if (!typeIsBound) {
+            System.out.printf("Please add %s to the list of bound classes, " +
+                                  "cause dynamic addition slows down perfomance%n", type.getName());
```

Comment:
I think a missing type in this private method should always indicate an error in the code. We never actually want the unmarshaller to be re-allocated to add types dynamically. So I wonder if we should use an assertion instead. Something like:

```java
assert classesToBeBound.contains(type) : type.getName();
```

Alternatively, do we really need to trigger our own error at all? What happens by default, if an `Unmarshaller` is asked to unmarshall a type that hasn't been passed to the `JAXBContext` instance? If there is some helpful exception then I think that would suffice.

Reply:
Catching original JAXBException and adding a little tip seems good
```java
java.lang.RuntimeException: Please, add class to classesToBeBound set if it is not known to this context
	at org.mapeditor.io.TMXMapReader.unmarshalClass(TMXMapReader.java:189)
	at org.mapeditor.io.TMXMapReader.buildMap(TMXMapReader.java:775)
	at org.mapeditor.io.TMXMapReader.unmarshal(TMXMapReader.java:839)
	at org.mapeditor.io.TMXMapReader.readMap(TMXMapReader.java:864)
	at org.mapeditor.io.TMXMapReader.readMap(TMXMapReader.java:881)
	at org.mapeditor.io.TestRead.lambda$main$0(TestRead.java:42)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: javax.xml.bind.JAXBException: org.mapeditor.core.Map is not known to this context
	at com.sun.xml.bind.v2.runtime.JAXBContextImpl.getBeanInfo(JAXBContextImpl.java:594)
	at com.sun.xml.bind.v2.runtime.unmarshaller.UnmarshallerImpl.getBeanInfo(UnmarshallerImpl.java:569)
	at com.sun.xml.bind.v2.runtime.unmarshaller.UnmarshallerImpl.unmarshal(UnmarshallerImpl.java:318)
	at org.mapeditor.io.TMXMapReader.unmarshalClass(TMXMapReader.java:187)
	... 9 more
```